NetKet/Output/json_output_writer.hpp

```diff
@@ -111,13 +111,14 @@ class JsonOutputWriter {
   // The first overload works for classes inheriting from AbstractMachine, the
   // second one for Eigen matrices.
   template <typename T>
-  void SaveState_Impl(std::ofstream& stream, const AbstractMachine<T>& state) {
-    state.Save(stream);
+  void SaveState_Impl(std::ofstream& stream,
+                      std::shared_ptr<AbstractMachine<T>> state) {
```

Comment:
I do not see why it would make sense to pass a shared pointer here. As above, nothing in `SaveState` retains the reference so it is safe to just pass a const reference to the machine.

The only reason to pass a shared pointer to a function (instead of just the underlying raw pointer or, equivalently, a reference) is if the functions requires shared ownership of the object. The output writer does not require this.

Reply:
Machine are handled and passed as shared pointers throughout the code, to avoid memory management problems in python. I am not sure I can pass a reference to the state instead of a pointer there because this is called from `VariationalMonteCarlo` where machine is stored as a shared pointer